### ISSUE TITLE
Shellcheck says that a function "references arguments" when it only checks if any argument has been passed

### ISSUE THREAD

**Issue by Ganton** [koalaman/shellcheck#2492]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2119 and SC2120
- My shellcheck version (`shellcheck --version` or "online"): 0.8.0 and online.
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

showHelp() {
    if [[ $# != 0 ]]; then
        echo "ERROR: A wrong quantity of arguments was used." >&2
    fi
    
    echo "Help would be shown there"
}

showHelp || exit 1
```

#### Here's what shellcheck currently says:
Line 3: showHelp() {
^-- [SC2120](https://github.com/koalaman/shellcheck/wiki/SC2120) (warning): showHelp references arguments, but none are ever passed.
 
Line 11: showHelp || exit 1
^-- [SC2119](https://github.com/koalaman/shellcheck/wiki/SC2119) (info): Use showHelp "$@" if function's $1 should mean script's $1.

#### Here's what I wanted or expected to see:

Shellcheck shouldn't say anything because no arguments have to be passed and it's useful to check if any argument is passed.
___

Thanks for Shellcheck, Vidar et al!

**Comment by rbairwell** [comment]
Just coming in to say that I'm encountering this as well.

* Rule ID: SC2120 and SC2119
* online

Example code:
```
#!/usr/bin/bash
fn() {
        [[ "$#" -lt 1 ]] || return 1
}

fn2() {
        [[ "$#" -eq 0 ]] || return 1
}

fn3() {
        [[ "$#" -ne 0 ]] && return 1
}

fn4() {
        if [[ "$#" -gt 0 ]]; then
                return 1
        fi
        return 0
}

fn
fn2
fn3
fn4
```

same sort of results as @Ganton , but I thought I'd just test a couple of different variations "just in case".

Only way around it is to include `# shellcheck disable=SC2120` just before the function definition (and not where the `if` statement is as I expected)

**Comment by koalaman** [comment]
There is some logic in excluding `$#`, but empirically it appears to be pretty rare to check it without then using `$@` or `$1` based on the result. Is this part of a larger example?

**Comment by Ganton** [comment]
> There is some logic in excluding `$#`, but empirically it appears to be pretty rare to check it without then using `$@` or `$1` based on the result. Is this part of a larger example?

If I have understood you correctly, yes, it is part of a bigger code, with several functions. In some of those functions: some arguments are used, in other functions: no argument is used and (as a kind of safety measure) it's checked if ─however─ an argument is passed. In those functions: `$#` is used (as seen in the examples), but neither `$@` nor `$1` are used.

**Comment by rbairwell** [comment]
> There is some logic in excluding `$#`, but empirically it appears to be pretty rare to check it without then using `$@` or `$1` based on the result. Is this part of a larger example?

Yes it is in my case as well. Basically, I've got functions which check that they have NOT been passed any arguments (for confirmation purposes) and will throw errors etc if `$#` is not zero/greater than 1 (i.e. I don't trust people using that bit of code to read documentation and/or I expect it to be mis-called) - seems to be identical to @Ganton's use case as well.
